File: LeetCode/76. Minimum Window Substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        // build hash map for last positions of characters in T, each character key maps to a value of deque
        unordered_map<char, deque<int>> last_pos;
        for (const auto & c : t) {
            if (last_pos.find(c) != last_pos.end()) last_pos[c].push_back(-1);
            else last_pos.emplace(c, deque<int>{-1});
        }
        
        // build hash map to indicate the remaining characters in T that have not been found
        unordered_map<char, int> found;
        for (auto it = last_pos.begin(); it != last_pos.end(); it++) found.emplace(it->first, it->second.size());
        
        // build a list of found positions in S of all characters in T
        vector<bool> found_pos(s.size(), false);
        
        string window = ""; // result window
        int j = -1; // position of the earliest character in S that matches T so that s[j:i+1] is a window
        bool all_found = false; // all characters in T have been found in S
        for (int i = 0; i < s.size(); i++) {
            if (last_pos.find(s[i]) != last_pos.end()) { // the character is in T
                // update found status
                if (j == -1) j = i;
                if (!all_found && found.find(s[i]) != found.end()) {
                    found[s[i]] -= 1;
                    if (!found[s[i]]) found.erase(s[i]);
                    if (found.empty()) {
                        all_found = true;
                        window = s.substr(j, i+1-j);
                    }
                }
                
                // update new last positions of s[i]
                int last = last_pos[s[i]].front();
                last_pos[s[i]].pop_front();
                last_pos[s[i]].push_back(i);
                
                // update found positions
                found_pos[i] = true;
                if (last >= 0) // guaranteed found[last] == True
                    found_pos[last] = false;
                    if (j == last) { // update window
                        while (!found_pos[j]) j++;
                        if (all_found && window.size() > i+1-j) window = s.substr(j, i+1-j);
                    }
            }
        }
        return window;
    }
};

File: LeetCode/76. Minimum Window Substring.py
from collections import deque

class Solution:
    def minWindow(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: str
        """
        # build hash map for last positions of characters in T, each character key maps to a value of deque
        last_pos = {}
        for c in t:
            if c in last_pos:
                last_pos[c].append(-1)
            else:
                last_pos[c] = deque([-1])
        
        # build hash map to indicate the remaining characters in T that have not been found
        found = {}
        for c in last_pos:
            found[c] = len(last_pos[c])
        
        # build a list of found positions in S of all characters in T
        found_pos = [False for _ in s]
        
        window = "" # result window
        j = -1 # position of the earliest character in S that matches T so that s[j:i+1] is a window
        all_found = False # all characters in T have been found in S
        for i in range(len(s)):
            if s[i] in last_pos: # the character is in T
                # update found status
                if j == -1: j = i
                if not all_found and s[i] in found:
                    found[s[i]] -= 1
                    if not found[s[i]]: del found[s[i]]
                    if not found:
                        all_found = True
                        window = s[j:i+1]
                
                # update new last positions of s[i]
                last = last_pos[s[i]][0]
                last_pos[s[i]].popleft()
                last_pos[s[i]].append(i)
                
                # update found positions
                found_pos[i] = True
                if last >= 0: # guaranteed found[last] == True
                    found_pos[last] = False
                    if j == last: # update window
                        while not found_pos[j]: j += 1
                        if all_found and len(window) > i-j+1: window = s[j:i+1]
        
        return window